bootstrap-app/src/cmr/bootstrap/data/fingerprint.clj

```diff
@@ -0,0 +1,101 @@
+  (:require
+    [clojure.core.async :as ca :refer [<!!]]
+    [cmr.bootstrap.embedded-system-helper :as helper]
+    [cmr.common.concepts :as concepts]
+    [cmr.common.log :refer (debug info warn error)]
+    [cmr.common.services.errors :as errors]
+    [cmr.metadata-db.data.concepts :as db]
+    [cmr.umm-spec.fingerprint-util :as fingerprint-util]))
+
+(defn- fingerprint-variable
+  "Update the fingerprint of the given variable if necessary."
+  [db provider variable]
+  (let [{:keys [concept-id revision-id deleted metadata]} variable]
+    (when (not deleted)
+      (let [old-fingerprint (get-in variable [:extra-fields :fingerprint])
+            new-fingerprint (fingerprint-util/get-variable-fingerprint metadata)]
+        (when (not= old-fingerprint new-fingerprint)
+          (db/save-concept db
+                           provider
+                           (-> variable
+                               (assoc :revision-id (inc revision-id))
+                               (assoc-in [:extra-fields :fingerprint] new-fingerprint)
+                               (dissoc :revision-date)))
+          (info (format "Updated fingerprint for concept-id: %s" concept-id)))))))
+
+(defn- fingerprint-variable-batch
+  "Update the fingerprints of variables in the given variable batch if necessary."
+  [db provider variable-batch]
+  (dorun
+   (map #(fingerprint-variable db provider %) variable-batch)))
+
+(def ^:private find-variables-sql-part1
+  "Defines the beginning part of the string to construct find variables sql statement."
+  (str "select a.* from metadata_db.cmr_variables a, "
+       "(select concept_id, max(revision_id) rid from metadata_db.cmr_variables "))
+
+(def ^:private find-variables-sql-part2
+  "Defines the ending part of the string to construct find variables sql statement."
+  (str "group by concept_id) b"
+       " where a.concept_id = b.concept_id and a.revision_id = b.rid and a.deleted = 0"))
+
+(defn- find-variables-sql
+  "Returns the sql statement to find the latest variable concept revisions that are not deleted."
+  ([]
+   (str find-variables-sql-part1 find-variables-sql-part2))
+  ([provider-id]
+   (format "%s where provider_id='%s' %s"
+           find-variables-sql-part1 provider-id find-variables-sql-part2)))
+
+(defn- fingerprint-by-provider
+  "Update the fingerprints of variables of the given provider if necessary."
+  [system provider]
+  (info "Updating fingerprints of variables for provider: " (:provider-id provider))
+  (let [db (helper/get-metadata-db-db system)
+        {:keys [provider-id]} provider
+        params {:concept-type :variable
+                :provider-id provider-id}
+        variable-batches (db/find-concepts-in-batches-with-stmt db
+                                                                provider
+                                                                params
+                                                                (find-variables-sql provider-id)
```

Comment:
Is `provider-id` sanitized? They way we construct the SQL using `format` seems like it could leave us open to SQL injection (I'm assuming `provider-id` is a POST parameter).

Reply:
The `provider-id` comes from POST body and it has to be validated before it gets here. See validation code at `cmr.bootstrap.data.fingerprint` line 68 and test at `cmr.system-int-test.bootstrap.fingerprint-test` line 159.